pilot/pkg/proxy/envoy/v2/grpc.go

```diff
@@ -0,0 +1,262 @@
+	}
+
+	err = con.send(resp)
+	if err != nil {
+		adsLog.Warnf("LDS: Send failure %s: %v", con.ConID, err)
+		recordSendError(ldsSendErrPushes, err)
+		return err
+	}
+
+	return nil
+}
+
+// Handle a gRPC CDS request, used with the 'ApiListener' style of requests.
+// The main difference is that the request includes Resources.
+func (s *DiscoveryServer) handleAPICDS(con *XdsConnection, req *xdsapi.DiscoveryRequest) bool {
+	if s.handleAck(con, req) {
+		return true
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     ClusterType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+
+	// gRPC doesn't currently support any of the APIs - returning just the expected EDS result.
+	// Since the code is relatively strict - we'll add info as needed.
+	for _, n := range req.ResourceNames {
+		hn, portn, err := net.SplitHostPort(n)
+		if err != nil {
+			adsLog.Warna("Failed to parse ", n, " ", err)
+			continue
+		}
+		rc := &xdsapi.Cluster{
+			Name: n,
+			ClusterDiscoveryType: &xdsapi.Cluster_Type{Type: xdsapi.Cluster_EDS},
+			EdsClusterConfig: &xdsapi.Cluster_EdsClusterConfig{
+				ServiceName: "outbound|" + portn + "||" + hn,
+				EdsConfig: &envoycore.ConfigSource{
+					ConfigSourceSpecifier: &envoycore.ConfigSource_Ads{
+						Ads: &envoycore.AggregatedConfigSource{},
+					},
+
+				},
+			},
+		}
+		rr := util.MessageToAny(rc)
+		resp.Resources = append(resp.Resources, rr)
+	}
+
+	err := con.send(resp)
+	if err != nil {
+		adsLog.Warnf("LDS: Send failure %s: %v", con.ConID, err)
+		recordSendError(ldsSendErrPushes, err)
+		return true
+	}
+	return true
+}
+
+// handleSplitRDS supports per-VIP routes, as used by GRPC.
+// This mode is indicated by using names containing full host:port instead of just port.
+// Returns true of the request is of this type.
+func (s *DiscoveryServer) handleSplitRDS(con *XdsConnection, req *xdsapi.DiscoveryRequest) bool {
+	for _, n := range req.ResourceNames {
+		if !strings.Contains(n, ":") {
+			return false // normal Istio RDS, on port
+		}
+	}
+
+	if s.handleAck(con, req) {
+		return true
+	}
+	push := s.globalPushContext()
+
+	// Currently this mode is only used by GRPC, to extract Cluster for the default
+	// route.
+	// Current GRPC is also expecting the default route to be prefix=="", while we generate "/"
+	// in normal response.
+	// TODO: add support for full route, make sure GRPC is fixed to support both
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     RouteType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+	for _, n := range req.ResourceNames {
+		hn, portn, err := net.SplitHostPort(n)
+		if err != nil {
+			adsLog.Warna("Failed to parse ", n, " ", err)
+			continue
+		}
+		port, err := strconv.Atoi(portn)
+		if err != nil {
+			adsLog.Warna("Failed to parse port ", n, " ", err)
+			continue
+		}
+		el := con.node.SidecarScope.GetEgressListenerForRDS(port, "")
+		vsvc := el.VirtualServices()
+		for _, s:= range vsvc {
```

Comment:
I am a bit confused what these virtual services are for? How do they map to gRPC?

Reply:
gRPC request for foo.com/my.service/method
We resolve foo.com to VIP and route name in LDS
RDS return the routes - including the VirtualService mappings for different URLs

GRPC doesn't parse the routes yet - but I hope it'll be high priority for them,